### PR TITLE
Enable Python 2.6 support

### PR DESCRIPTION
**Pull request by bcipolli**
This partially addresses #357.

There are a few issues preventing the setup script, code, and tests from running in Python 2.6:
* Trailing whitespace in `version.py` led to a setup syntax error.
* `"format_str{}".format(str)` is illegal in Python 2.6; use `"format_str{0}".format(str)`
* Dictionary comprehensions are NYI in Python 2.6; use `dict([list_comprehension])`
* Many `assert_X` functions are unavailable in Python 2.6; convert the few we were using to `assert_true`
* Context managers for `ZipFile`, `TarFile` don't exist in Python 2.6; wrap with `contextlib`
* Context manager for `nose.assert_raises` doesn't exist in Python 2.6; use our custom `assert_raises_regexp`, which was more appropriate anyway (since we wanted to check the error message of the error anyway).

I've also changed the `.travis.yml` file to contain a flow like `scikit-learn`.  The following three scenarios are now enabled:
* Python 2.6 under Anaconda install
* Python 2.7 using system packages
* Python 2.7 under neurodebian.

Tests for all three scenarios are working (or were working; seems the PGP key for neurodebian is choking at this moment, out of the blue)...

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,23 +1,20 @@
...
-    - "2.7"
-
 virtualenv:
   system_site_packages: true
 
 env:
-  - NEURODEBIAN=0
-  - NEURODEBIAN=1
-
-before_install:
```
Comment (bcipolli) [inline .travis.yml]: This all got moved into the `install.sh` file.

**Comment by lesteve** [comment]
Thanks a for for this, this is super useful!

> Tests for all three scenarios are working (or were working; seems the PGP key for neurodebian is choking at this moment, out of the blue)...

Yeah this is quite annoying when that happens, not sure whether there is a way to work around it. I restarted the job on Travis and everything seems fine now.

File: continuous_integration/install.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+else
+    echo "Unrecognized distribution ($DISTRIB); cannot setup travis environment."
+    exit 1
+fi
+
+python --version
+python -c "import numpy; print('numpy %s' % numpy.__version__)"
+python -c "import scipy; print('scipy %s' % scipy.__version__)"
+python -c "import sklearn; print ('sklearn %s' % sklearn.__version__)"
```
Comment (lesteve) [inline continuous_integration/install.sh]: It'd be quite nice to add other dependencies versions here. I am thinking matplotlib and nibabel.

File: nilearn/plotting/glass_brain.py [hunk 3]
```diff
@@ -96,26 +96,26 @@ def _get_json_and_transform(direction):
 
     dirname = os.path.dirname(os.path.abspath(__file__))
     dirname = os.path.join(dirname, 'glass_brain_files')
-    direction_to_filename = {
-        direction: os.path.join(
+    direction_to_filename = dict([
+        (_direction, os.path.join(
```
Comment (lesteve) [inline nilearn/plotting/glass_brain.py]: Nice catch, it took me a little while to figure out why you were renaming the list comprehension variables ... (answer: shadowing the _get_json_and_transform 'direction' parameter)

**Comment by GaelVaroquaux** [comment]
Thanks for the PR. This is absolutely great!

> * Many `assert_X` functions are unavailable in Python 2.6; convert the few we were using to `assert_true`

Have you check in sklearn.utils.testing. I believe that we have
backported everything that was necessary there.

> * Context manager for `nose.assert_raises` doesn't exist in Python 2.6; use our custom `assert_raises_regexp`, which was more appropriate anyway (since we wanted to check the error message of the error anyway).

Same thing, there is a fairly developed framework for testing assertions
in scikit-learn. We should rely on it, as these things are actually quite
tricky to get right.

File: .travis.yml [hunk 4]
```diff
@@ -1,23 +1,20 @@
...
-    - wget -O- http://neuro.debian.net/lists/precise.us-nh.libre | sudo tee /etc/apt/sources.list.d/neurodebian.sources.list
-    - sudo apt-key adv --recv-keys --keyserver pgp.mit.edu 2649A5A9
-    - if [ $NEURODEBIAN -eq 1 ]; then sudo apt-get update -qq; fi
-    - sudo apt-get install -qq python-scipy python-nose python-nibabel python-sklearn
+  matrix:
+    - DISTRIB="ubuntu" PYTHON_VERSION="2.7"
+      COVERAGE="false"
+    - DISTRIB="neurodebian" PYTHON_VERSION="2.7"
+      COVERAGE="false"
```
Comment (lesteve) [inline .travis.yml]: I don't think the COVERAGE environment variable is used anywhere in the scripts or travis.yml. Also coverage is enabled in setup.cfg.

Which reminds me It'd be quite nice to setup coversall one day similarly to what scikit-learn is doing.

File: nilearn/_utils/niimg_conversions.py [hunk 5]
```diff
@@ -265,7 +265,7 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False,
         else:
             nii_str = "image #" + str(index)
         if verbose > 0:
-            print "Concatenating {}/{}: {}".format(index + 1, sum(lengths),
+            print "Concatenating {0}/{1}: {2}".format(index + 1, sum(lengths),
                                                    nii_str)
```
Comment (lesteve) [inline nilearn/_utils/niimg_conversions.py]: bonus points if you fix this PEP8 warning, while you are touching the line above.

**Comment by lesteve** [comment]
> Same thing, there is a fairly developed framework for testing assertions in scikit-learn. We should rely on it, as these things are actually quite tricky to get right.

For nilearn._utils.testing.assert_raises_regexp, I actually copied the code from scikit-learn. The main reason I did that is [there](https://github.com/nilearn/nilearn/commit/a0c7693e0325dea9c7ce04a8512936eede881d32). Also this was added only in scikit-learn 0.15, so we can not rely only on scikit-learn for this functionality since we need to support much older versions.

**Comment by GaelVaroquaux** [comment]
> Also this was added only in scikit-learn 0.15, so we can not rely only
> on scikit-learn for this functionality since we need to support much
> older versions.

OK. Good reason.

**Comment by bcipolli** [comment]
>For nilearn._utils.testing.assert_raises_regexp, I actually copied the code from scikit-learn. The main reason I did that is there. Also this was added only in scikit-learn 0.15, so we can not rely only on scikit-learn for this functionality since we need to support much older versions.

Good point, but supporting scikit-learn < 0.15 for users is different than supporting it for developers.  I think that, to avoid copying code and to use the full scikit-learn test functionality, it would be reasonable to require developers to use scikit-learn >= 0.15 (and show a meaningful error if they are not).

That doesn't seem to challenging to do... I'd be willing to make that happen if others are on-board with it.

**Comment by lesteve** [comment]
> Good point, but supporting scikit-learn < 0.15 for users is different than supporting it for developers. 

Not sure where you would draw the line and which mechanism you would use. The thing is that Travis need to run the tests with old versions of scikit-learn so I don't really see how we can avoid providing an assert_raises_regexp inside nilearn.

Also as you might have seen from the link in my previous message there is a small but very nasty bug that is only fixed in scikit-learn master: assert_raises_regexp provided by scikit-learn is using match rather than search to check the error message. As far as I was concerned that was the last straw that broke the camel's back of my attempts at reusing scikit-learn assert_raises regexp.

**Comment by bcipolli** [comment]
>The thing is that Travis need to run the tests with old versions of scikit-learn so I don't really see how we can avoid providing an assert_raises_regexp inside nilearn.

Good point.  OK then! :)

File: nilearn/plotting/glass_brain.py (hunk 3 above)
Comment (bcipolli) [inline nilearn/plotting/glass_brain.py]: Yup!  And nice catch; I should have added a comment to the PR to point out that change...

File: .travis.yml (hunk 4 above)
Comment (bcipolli) [inline .travis.yml]: I will remove these and add an issue for setting up `coversall`.  It's along the lines of things I'm able and happy to contribute.

Comment (GaelVaroquaux) [inline .travis.yml]: > I will remove these and add an issue for setting up coversall. It's along the
> lines of things I'm able and happy to contribute.

It's really great to have you on board, Ben!

**Comment by bcipolli** [comment]
Made the few cleanup / reporting tweaks; and made no further changes to the tests.  Let me know if there's anything else to do here; otherwise, I'll focus my time on completing Python 3.x support.

Thanks!

**Comment by lesteve** [comment]
This looks good to me, merging, Thanks a lot for that!

**Pull request merged**